File: airbyte-integrations/connectors/destination-bigquery/src/test/java/io/airbyte/integrations/destination/bigquery/BigQueryDestinationTest.java
/*
 * Copyright (c) 2021 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.destination.bigquery;

import static org.junit.jupiter.api.Assertions.*;

import com.fasterxml.jackson.databind.JsonNode;
import com.google.common.collect.ImmutableMap;
import io.airbyte.commons.json.Jsons;
import java.util.stream.Stream;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.*;

public class BigQueryDestinationTest {

  private ImmutableMap.Builder<Object, Object> configMapBuilder;

  @BeforeEach
  public void init() {
    configMapBuilder = ImmutableMap.builder()
        .put(BigQueryConsts.CONFIG_CREDS, "test_secret")
        .put(BigQueryConsts.CONFIG_DATASET_LOCATION, "US");
  }

  public static Stream<Arguments> validBigQueryIdProvider() {
    return Stream.of(
        Arguments.arguments("my-project", "my_dataset", "my_dataset"),
        Arguments.arguments("my-project", "my-project:my_dataset", "my_dataset"));
  }

  @ParameterizedTest
  @MethodSource("validBigQueryIdProvider")
  public void testGetDatasetIdSuccess(String projectId, String datasetId, String expected) throws Exception {
    JsonNode config = Jsons.jsonNode(configMapBuilder
        .put(BigQueryConsts.CONFIG_PROJECT_ID, projectId)
        .put(BigQueryConsts.CONFIG_DATASET_ID, datasetId)
        .build());

    String actual = BigQueryDestination.getDatasetId(config);

    assertEquals(expected, actual);
  }

  public static Stream<Arguments> invalidBigQueryIdProvider() {
    return Stream.of(
        Arguments.arguments("my-project", ":my_dataset", "BigQuery Dataset ID format must match '[project-id:]dataset_id': :my_dataset"),
        Arguments.arguments("my-project", "your-project:my_dataset",
            "Project ID included in Dataset ID must match Project ID field's value: Project ID is my-project, but you specified your-project in Dataset ID"));
  }

  @ParameterizedTest
  @MethodSource("invalidBigQueryIdProvider")
  public void testGetDatasetIdFail(String projectId, String datasetId, String expected) throws Exception {
    JsonNode config = Jsons.jsonNode(configMapBuilder
        .put(BigQueryConsts.CONFIG_PROJECT_ID, projectId)
        .put(BigQueryConsts.CONFIG_DATASET_ID, datasetId)
        .build());

    Exception exception = assertThrows(IllegalArgumentException.class, () -> BigQueryDestination.getDatasetId(config));

    assertEquals(expected, exception.getMessage());
  }

}

File: airbyte-integrations/connectors/destination-bigquery/src/test/kotlin/io/airbyte/integrations/destination/bigquery/BigqueryDestinationTest.kt
/*
 * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.destination.bigquery

import io.airbyte.commons.exceptions.ConfigErrorException
import io.airbyte.integrations.base.destination.typing_deduping.ColumnId
import io.airbyte.integrations.base.destination.typing_deduping.ParsedCatalog
import io.airbyte.integrations.base.destination.typing_deduping.StreamConfig
import io.airbyte.integrations.base.destination.typing_deduping.StreamId
import io.airbyte.integrations.base.destination.typing_deduping.Struct
import io.airbyte.protocol.models.v0.DestinationSyncMode
import io.airbyte.protocol.models.v0.SyncMode
import java.util.Optional
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class BigqueryDestinationTest {

    @Test
    fun throwOnBadStream() {
        val exception =
            Assertions.assertThrows(ConfigErrorException::class.java) {
                BigQueryDestination.throwIfAnyUnsupportedPrimaryKeys(
                    false,
                    buildCatalog(DestinationSyncMode.APPEND_DEDUP, 2)
                )
            }
        Assertions.assertEquals(
            """
                JSON-typed columns are not currently supported in primary keys.
                public.users_1: weird_column_1
                public.users_2: weird_column_2
            """.trimIndent(),
            exception.message
        )
    }

    @Test
    fun noThrowOnNonDedup() {
        Assertions.assertDoesNotThrow {
            BigQueryDestination.throwIfAnyUnsupportedPrimaryKeys(
                false,
                buildCatalog(DestinationSyncMode.APPEND)
            )
        }
    }

    @Test
    fun noThrowWhenTypingDedupingDisabled() {
        Assertions.assertDoesNotThrow {
            BigQueryDestination.throwIfAnyUnsupportedPrimaryKeys(
                true,
                buildCatalog(DestinationSyncMode.APPEND_DEDUP)
            )
        }
    }

    private fun buildCatalog(
        destinationSyncMode: DestinationSyncMode,
        numStreams: Int = 1
    ): ParsedCatalog {
        val streams =
            (1..numStreams)
                .map {
                    val pkColumn = ColumnId("", "weird_column_$it", "")
                    StreamConfig(
                        StreamId("", "", "", "", "public", "users_$it"),
                        SyncMode.INCREMENTAL,
                        destinationSyncMode,
                        listOf(pkColumn),
                        Optional.empty(),
                        linkedMapOf(pkColumn to Struct(linkedMapOf()))
                    )
                }
                .toList()
        return ParsedCatalog(streams)
    }
}
